### PR TITLE
Improved types

### PR DESCRIPTION
**Pull request by wtrocki**
Improvements for typescript interfaces. Tests for this work are in sync client demo.

### CODE REVIEW

File: fh-sync-js.d.ts [hunk 1]
```diff
@@ -1,329 +1,424 @@
...
+   * @interface NotifyCallback
+   */
+  interface NotifyCallback {
+    (data: NotificationData): any
+  }
+
+  /**
+   * Initialize the client data sync service.
+   *
+   * @param {Object} options
```
Comment (JameelB) [inline fh-sync-js.d.ts]: cloudUrl missing from the param description?

File: fh-sync-js.d.ts [hunk 2]
```diff
@@ -1,329 +1,424 @@
...
+    storage_strategy?: "html5-filesystem" | "dom" | "webkit-sqlite" | "indexed-db" | any;
+
+    /**
+     * Amount of space to request from the HTML5 filesystem API when running in browser
+     * Optional. Default: 50 * 1024 * 1024
+     */
+    file_system_quota?: number,
+
+    /**
+     * iOS only. If set to true, the file will be backed by iCloud.
```
Comment (JameelB) [inline fh-sync-js.d.ts]: no default value given on description

File: fh-sync-js.d.ts [hunk 3]
```diff
@@ -1,329 +1,424 @@
...
+   * @param {Boolean} [options.sync_active=true] - Is the background synchronization with the cloud currently active. If this is set to false, the synchronization loop will not start automatically. You need to call startSync to start the synchronization loop.
+   * @param {String} [options.storage_strategy=html5_filesystem] - Storage strategy to use for the underlying client storage framework Lawnchair. Valid values include 'dom', 'html5-filesystem', 'webkit-sqlite', 'indexed-db'. Multiple values can be specified as an array and the first valid storage option will be used. If the app is running on Titanium, the only support value is 'titanium'.
+   * @param {Number} [options.file_system_quota=52428800] - Amount of space to request from the HTML5 filesystem API when running in browser
+   * @param {Boolean} [options.icloud_backup=false] - iOS only. If set to true, the file will be backed by iCloud.
+   */
+  function init(options: SyncOptions): void;
+
+  /**
+   * Register a callback function to be invoked when the sync service has notifications to communicate to the client.
+   *
```
Comment (JameelB) [inline fh-sync-js.d.ts]: missing dataset_id param description

File: fh-sync-js.d.ts [hunk 4]
```diff
@@ -1,329 +1,424 @@
...
-         */
-        function setStorageAdapter(handler: (dataset_id:string, isSave: boolean, cb: any) => void);
-
-        /**
-         * Allows to override default hashing method
-         *
-         * @param handler - function that wraps underlying hashing method
-         */
-        function setHashMethod(method: () => any);
+
```
Comment (wtrocki) [inline fh-sync-js.d.ts:7]: I just fixed strange 4 line formatting here.

File: fh-sync-js.d.ts [hunk 5]
```diff
@@ -1,329 +1,424 @@
...
+    message?: string;
+    code: string;
+  }
+
+  /**
+   * Interface for the options object provided to the init function.
+   *
+   * @interface SyncOptions
+   */
+  interface SyncOptions {
```
Comment (wtrocki) [inline fh-sync-js.d.ts:24]: Added comments

File: fh-sync-js.d.ts [hunk 6]
```diff
@@ -1,329 +1,424 @@
...
+   * Register a callback function to be invoked when the sync service has notifications to communicate to the client.
+   *
+   * @param {Function} callback
+   */
+  function notify(dataset_id: string, callback: NotifyCallback): void;
+  /**
+   * Put a dataset under the management of the sync service.
+   *
+   * @param {String} datasetId
+   * @param {Object} options
```
Comment (JameelB) [inline fh-sync-js.d.ts]: missing cloudUrl param description

File: fh-sync-js.d.ts [hunk 7]
```diff
@@ -1,329 +1,424 @@
...
+  function loadDataSet(datasetId: string, success: (dataset: any) => void, failure: () => void): void;
+
+  /**
+   * @param {String} datasetId
+   * @param {Function} callback
+   */
+  function clearCache(datasetId: string, callback?: () => void): void;
+
+  /**
+   * Sets cloud call handler for sync. Required to make any sync requests to the server side (sync cloud)
```
Comment (JameelB) [inline fh-sync-js.d.ts]: no param description

File: fh-sync-js.d.ts [hunk 8]
```diff
@@ -1,329 +1,424 @@
...
+  /**
+   * Allows to override default storage adapter
+   *
+   * @param handler - function that wraps underlying storage solution
+   */
+  function setStorageAdapter(handler: (dataset_id: string, isSave: boolean, cb: any) => void): void;
+  /**
+   * Allows to override default hashing method
+   *
+   * @param handler - function that wraps underlying hashing method
```
Comment (JameelB) [inline fh-sync-js.d.ts]: wrong param name

**Review by JameelB — APPROVED** [review]
small comments

File: fh-sync-js.d.ts [hunk 9]
```diff
@@ -1,329 +1,429 @@
...
+   * Register a callback function to be invoked when the sync service has notifications to communicate to the client.
+   * 
+   * @param dataset_id - dataset identifier
+   * @param {Function} callback
+   */
+  function notify(dataset_id: string, callback: NotifyCallback): void;
+  /**
+   * Put a dataset under the management of the sync service.
+   *
+   * @param {String} datasetId
```
Comment (austincunningham) [inline fh-sync-js.d.ts:192]: needs a comment to define datasetId , took me a long time to determine what datasetId was referring to .

**Pull request merged**